design-proposals/cpu-hotplug.md

```diff
@@ -0,0 +1,163 @@
+# Overview
+This proposal is about adding CPU hotplug support to KubeVirt.
+CPU hotplug will allow users to dynamically add/remove CPU resources to/from running VMs.
+While this design is focused on CPU hotplug, it aims to provide a generic API design for resource hotplug that will be reusable for things like memory.
+
+## Motivation
+CPU hotplug is now a common virtualization feature, but KubeVirt lacks support for it.
+Live vertical scaling of VM resources in general can be very useful in a variety of scenarios.
+
+## Goals
+- Hotplug and hot-unplug of CPU resources from VMs should be achievable by editing the VM object itself (no sub-resources)
+- Underlying technology used for hotplug should be independent of the VM API, meaning we should be able to swap out how hotplug is technically performed in the future without impacting API compatibility
+- Implementation should be achievable today, with Kubernetes APIs that are at least in beta. Unfortunately, at the time of writing, the Kubernetes vertical pod scaling API is only alpha
+
+## Non Goals
+- Revising existing hotplug mechanisms like Volume hotplug
+- Dynamically applying affinity rules to VMIs
+
+## Definition of Users
+A user with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+## User Stories
+As a VM user, I would like to dynamically increase or decrease the number of CPUs used by a VM without requiring the VM to restart.
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+In VM objects, the existing CPU `sockets` field will now be dynamic, provided that it has been declared as such.
+Declaring the field as dynamic will be done by adding a `cpu` entry to a new section under the VM spec, called `liveUpdateFeatures`.
```

Comment:
For the network interface hotplug, I suggested to use the existing domain resources requests (or/and limits) with a dedicated resource name. The semantics of "requests" is for assuring reservation and "limits" for placing an upper limit.

Is there a reason not to re-use the "map" of resources for the same?

Reply:
I see the map of `liveUpdateFeatures` to be inconsistent with ` resources requests`, in the future `liveUpdateFeatures` may contain not just resources thinks like nodeSelector, etc...

Reply:
Well, I do not understand the grouping then.
Subject domains are distributed in the spec, but this one seems to try and group "live-update" and de facto duplicates other subject domains trees.

The parts which are about reservation or limitation of resources, can re-use the `Domain.Resources` structure. When the scope changes and requires a more complex coordination, new things could be created.

Are the existing resources discussed in this proposal not a fit for the domain resource semantics?
Is there an example of something that is not a fit? I want to understand better the `resources` limitation in this regard.

Reply:
Normally, changes made to the VMI template will not take effect immediately, but only on a creation of a new VMI.
The new `vm.spec.liveUpdateFeatures` should tell which changes to the vmi template will propagate immediately.
Having this I'm mind, I don't see a clear connection to the resources struct.   

> Are the existing resources discussed in this proposal not a fit for the domain resource semantics?
> Is there an example of something that is not a fit? I want to understand better the resources limitation in this regard.

This is not part of this proposal, however, it was discussed to add something like `Affinity` will be added in the future to dynamically control the node affinity of the VMI. 
Additionally, the Memory struct which will be added in the future, could preallocate the hot-plugged memory, so it could have something like a `MaxQuantity` field in it.

I don't see a reason why we need to force the liveUpdateFeatures to use the resource struct.

Reply:
> Normally, changes made to the VMI template will not take effect immediately, but only on a creation of a new VMI. The new `vm.spec.liveUpdateFeatures` should tell which changes to the vmi template will propagate immediately. Having this in mind, I don't see a clear connection to the resources struct.

Indeed, changes to the VM spec template are currently never passed on to the underlying VMI.
Storage hotplug have and not network hotplug are using a command queue to propagate such changes.

Is this proposal is suggesting a new way to propagate live changes?

> I don't see why we need to force the liveUpdateFeatures to use the resource struct.

That was not my intention.
The way storage & network hotplug are propagated from a VM to a VMI is through a queue on the VM status.
Users can only use the virt-api to update live changes, not directly on the VM. I.e. changes on the VM directly are never "live".

The usage of the resource generic structure is to pass in various key-value that represent resources of the domain, relevant for both immutable and mutable resources. The ones that are live updatable, are processed by the VMI, the others are left for the next boot. If there are existing fields for the resource in other places (e.g. disks, volumes, interfaces, cpu count), no need to add to the resource key/value anything.

> > Are the existing resources discussed in this proposal not a fit for the domain resource semantics?
> > Is there an example of something that is not a fit? I want to understand better the resources limitation in this regard.
> 
> This is not part of this proposal, however, it was discussed to add something like `Affinity`, which will be added in the future to control the node affinity of the VMI dynamically. Additionally, the Memory struct which will be added in the future, could preallocate the hot-plugged memory, so it could have something like a `MaxQuantity` field in it.
> 
> I don't see why we need to force the liveUpdateFeatures to use the resource struct.

Reply:
I honestly like this proposal as is, the introduction of the `liveUpdateFeatures` field makes it straightforward to understand which resources are live updatable in a running VM.

To me, it just makes it easier to better understand and use the hotplug capabilities of the VM.

Reply:
This suggestion is not straightforward to me, because it introduces two places for the same information to exist. At least, this is not a simple solution, considering that users just want to request changes to a running VM.
But the real interesting feedback is from the feature stakeholders, which should confirm if this fits their needs or not.

But my main concern is about not having a common consistent method to do live updates.
IMO this needs to be discussed in a cross SIG forum and once agreed upon on the API, to convert all to the same method.

Reply:
>I honestly like this proposal as is, the introduction of the liveUpdateFeatures field makes it straightforward to understand which resources are live updatable in a running VM.

It seems that it does more than that, it enables setting an upper limit for CPU sockets inside the guest.
Per my understanding this can be achieved through `resource.limits` or `spec.domain.cpu`.

As for today you have two places to set the CPU resource, adding another place make the API more cumbersome and worse experience IMO.